### ISSUE TITLE
[noteout] doesn't send note off msg

### ISSUE THREAD

**Issue by mariobuoninfante** [pure-data/pure-data#796]
Hi,

Don't think this is a bug, but I just realized that [noteout] doesn't send _note off_ msg, and instead _note on_ with velocity 0.
I know you can use [midiout] to achieve that, but I was wondering whether it would be possible to add the possibility to decide what to send when the object receives a velocity value of 0.

Cheers,
Mario

**Comment by mariobuoninfante** [comment]
this just popped up in my mind,
maybe setting the velocity to a negative value could be a suitable option to allow sending note off messages?
also cause there are piece of hardware that uses note off velocity values (!=0) to affect sound parameters, ie staccato and stuff like that

**Comment by Spacechild1** [comment]
> what to send when the object receives a velocity value of 0.

when you send a velocity of 0, you don't need to send a note-off message, because you loose the velocity part.

For symmetry, I guess we could simply add [noteoffin] and [noteoffout] objects, but I'm not sure it's really worth it...

**Comment by porres** [comment]
> Don't think this is a bug, but I just realized that [noteout] doesn't send note off msg, and instead note on with velocity 0.

it's not a bug, a note on with velocity 0 is the definition of a note off message in the MIDI world

**Comment by Spacechild1** [comment]
> a note on with velocity 0 is the definition of a note off message in the MIDI world

No, a note-on message with velocity 0 is not the same as a note-off message. The MIDI standard only says that the former must be treated the same as the latter.

**Comment by mariobuoninfante** [comment]
yap, I know sending a note on with vel 0 is not an issue :)
mine was a statement, like _I know it is not a bug, but...._
@Spacechild1 
not sure I got what you mean with
> when you send a velocity of 0 to [noteout] to denote a note-off, you don't need to send an actual note-off message, because the velocity part has obviously been lost.
What I was trying to say there is basically that it would be great to be able to send note off messages using [noteout] (since the name doesn't imply note ON or OFF).

**Comment by porres** [comment]
I think you're talking about 'release velocity' then? It's not the same as 'note off'

**Comment by porres** [comment]
The only way to emulate release velocity in vanilla is with [midiout] I guess.

Anyway, objects in 'cyclone' and 'else' can deal with release velocity.

See cyclone/xnoteout and  else/note.out

**Comment by porres** [comment]
I guess you could rename the issue to "feature: make [noteout] send release velocity". 

Maybe a new inlet/outlet to [notein]/[noteout] could do the trick (as in cyclone/xnoteout) In else I put a '-rel' flag to add this extra parameter, also an option.

**Comment by Spacechild1** [comment]
> I think you're talking about 'release velocity' then? It's not the same as 'note off'

note-off messages are status byte 128-143, note-on messages are 144-159. A note-on message with velocity 0 is treated like a note-off message, but it *is not* a note-off message

**Comment by Spacechild1** [comment]
having an (optional) additional inlet/outlet for [notein] / [noteout] is not a bad idea, though.

**Comment by mariobuoninfante** [comment]
basically lie @Spacechild1 was saying you have:
note off: 
byte1 = 128-143 (channel 1-16)
byte2 = pitch
byte3 = velocity

then you have note on:
byte1 = 144-159 (channel 1-16)
byte2 = pitch
byte3 = velocity

my request (and I agree I should change the title to reflect that is a request) was to add a way to send note off _bytes_
this because the 3rd byte (velocity) of a note off message is used by some instruments.
I like the idea of the extra inlet/outlet, why do you say _optional_? are you thinking about anything specific?
when I read that about the extra inlet/outlet, I though about something like sending/receiving a 0 when it's a note off, and 1 when is a note on.

**Comment by Spacechild1** [comment]
since we have objects for all other MIDI message types (note-on, cc, pitchbend, aftertouch, channel aftertouch, program change), why not simply add [noteoffin] and [noteoffout] to make it symmetric?

**Comment by mariobuoninfante** [comment]
I'd tend to agree with, it sounds ok on paper. 
I'm just a bit concerned about the possible workflow there.
in case you have 1 single object that can handle both _note on_ and _note off,_ you could for example simply flip a bit and turn a _note on_ in a _note off_ and the other way around.
In case you have 2 separate objects, then it seems to me as if it gets over-complicated.
You start dealing with routing notes, not rocket-science I know, but...

**Comment by porres** [comment]
> basically like @Spacechild1 was saying, you have:
> 
> note off
> byte1 = 128-143 (channel 1-16)
> byte2 = 0-127 (pitch)
> byte3 = 0-127 (velocity)
> 
> then you have
> 
> note on
> byte1 = 144-159 (channel 1-16)
> byte2 = 0-127 (pitch)
> byte3 = 0-127 (velocity)

Yup, I'm aware, having worked with cyclone and my own design of an object, I see. I guess I might be confusing the terms as I always thought of 'note off velocity' as simply 'release velocity'. But anyway, I see what you want now and I think it'd be great to have that built into Pd Vanilla.

Now, as for the design. Having a better look, simply an extra in/outlet wouldn't do it. So I guess a flag could be the case, or just a new object. I went for the 'flag' option in my design.

But anyway, bulding an abstraction over [midiin]/[midiout] wouldn't be too hard to achieve this for a quick/urgent solution.

**Comment by porres** [comment]
> when I read that about the extra inlet/outlet, I though about something like sending/receiving a 0 when it's a note off, and 1 when is a note on.

yeah, that's what it'd do, but then you'd stop from having a note on velocity 0 for a note off message, and this will break EVERY patch. So a flag seems like the only option here to keep compatibility.

**Comment by mariobuoninfante** [comment]
yap, you're right about the compatibility, definitely.
Actually thinking about [notein] in particular, if there's is then an extra inlet, unless you explicitly deal with that it would be super easy to mess up noteons and noteoffs.
this makes me think the separate object would probably be the best way, and yes the abstraction solution (both with [midiin]/[midiout] or the new object) is always a solution
